mill-universal-packager/src/io/github/hoangmaihuy/mill/packager/universal/Archiver.scala

```diff
@@ -0,0 +1,179 @@
+    */
+  def mkTarball(
+    input: Path,
+    output: Path,
+    wrappedIn: Option[SubPath]
+  )(
+    compress: OutputStream => CompressorOutputStream
+  ): Unit = {
+    Using.Manager { use =>
+      val fos = use(new FileOutputStream(output.toIO))
+      val bos = use(new BufferedOutputStream(fos))
+      addPathsToTarball(input, compress(bos), wrappedIn)
+    }
+  }
+
+  /** make a compressed tarball file with the compression function from the given mappings
+    * @param mappings
+    *   the file mappings as input
+    * @param out
+    *   the output path
+    * @param wrappedIn
+    *   the optional sub path to wrap the contents in the tarball file. eg: for mapping (/tmp/xyz/1.conf -> conf/1.conf)
+    *   and wrappedIn = archived/simple-app-v0.0.1-SNAPSHOT, it will produce
+    *   archived/simple-app-v0.0.1-SNAPSHOT/conf/1.conf in the tarball file
+    * @param compress
+    *   the compression function
+    */
+  def mkTarball(
+    mappings: Seq[(Path, SubPath)],
+    output: Path,
+    wrappedIn: Option[SubPath]
+  )(
+    compress: OutputStream => CompressorOutputStream
+  ): Unit =
+    Using.Manager { use =>
+      val fos = use(new FileOutputStream(output.toIO))
+      val bos = use(new BufferedOutputStream(fos))
+      val tos = use(new TarArchiveOutputStream(compress(bos)))
+      mappings.foreach { case (from, to) => addPathToTar(from, to, tos, wrappedIn) }
+    }
+
+  // recursively add a Path which can be a file or a directory to the given TarArchiveOutputStream.
+  // if the path is a single file, just add it to the output stream,
+  // if the path is a directory, add all of the contents under the path to the output stream
+  private def addPathsToTarball(path: Path, outputStream: OutputStream, wrappedIn: Option[SubPath]): Unit = {
+    Using.resource(new TarArchiveOutputStream(outputStream)) { tarArchive =>
+      path match {
+        case _ if os.isDir(path) =>
+          os.walk(path).foreach { p =>
+            addPathToTar(
+              path = p,
+              target = p.subRelativeTo(path),
+              tarArchive = tarArchive,
+              wrappedIn = wrappedIn
+            )
+          }
+        case _ if os.isFile(path) =>
+          addPathToTar(
+            path = path,
+            target = path.subRelativeTo(path / os.up),
+            tarArchive = tarArchive,
+            wrappedIn = wrappedIn
+          )
+        // this should not happen
+        case _ =>
+      }
+    }
+  }
+
+  // add single Path which can be a file or a directory to the given TarArchiveOutputStream, not recursively
+  private def addPathToTar(
+    path: Path,
+    target: SubPath,
+    tarArchive: TarArchiveOutputStream,
+    wrappedIn: Option[SubPath]
+  ): Unit = {
+    val file: File = path.toIO
+    // Create entry name relative to parent file path
+    val entryPath = wrappedIn.map(_ / target).getOrElse(target)
+    val tarEntry = new TarArchiveEntry(file, entryPath.toString)
+    SOURCE_DATE_EPOCH.foreach { d =>
```

Comment:
Do we need validation for epoch?

Reply:
i think just filtering out the negative value should be enough, as the building server might have incorrect time.
```scala
  private val SOURCE_DATE_EPOCH: Option[Long] = ctx.env.get("SOURCE_DATE_EPOCH").flatMap(_.toLongOption).filter(_ > 0L)
```